benchmark/cmd/querier/main.go

```diff
@@ -86,106 +115,113 @@ func main() {
+			Query:       query,
+			queryURL:    queryURL,
+			MinDuration: math.MaxInt64,
+		})
 	}
 
-	queryURL := fmt.Sprintf(queryURLTemplate, url.QueryEscape(query))
-
-	level.Info(logger).Log("Starting testing query", query)
-
-	// Run queries concurrently.
+	summaryStart := time.Now()
 	var wg sync.WaitGroup
-	for i := 0; i < concurrentReqs; i++ {
+	for i := 0; i < *fConcurrentRequests; i++ {
 		wg.Add(1)
-		go performQuery(logger, res, queryURL, timeout, &wg)
+		go performQuery(logger, t, *fQueryTotalTime, &wg)
 	}
-
 	wg.Wait()
+	t.TotalDuration = time.Since(summaryStart)
 
-	level.Info(logger).Log("Finished testing query", query)
+	// launch server if needed
+	if *fServer {
+		resultsBytes, err := json.Marshal(t.results)
+		if err != nil {
+			level.Error(logger).Log("err", err)
+			return
+		}
+
+		http.HandleFunc("/results", func(w http.ResponseWriter, req *http.Request) {
+			if _, err := w.Write(resultsBytes); err != nil {
+				level.Error(logger).Log("err", err)
+				w.WriteHeader(http.StatusInternalServerError)
+			}
+		})
+
+		level.Info(logger).Log("msg", "Serving results")
+		if err := http.ListenAndServe(":8080", nil); err != nil {
+			level.Error(logger).Log("msg", "could not start http server")
+		}
+	}
 
-	return res
+	err := printResults(t)
+	if err != nil {
+		level.Error(logger).Log("err", err)
+	}
 }
 
-func performQuery(logger log.Logger, resultStorage *QueryResult, query string, timeout time.Duration, wg *sync.WaitGroup) {
+func performQuery(logger log.Logger, totals *totals, timeout time.Duration, wg *sync.WaitGroup) {
 	defer wg.Done()
 
-	var totalDuration time.Duration
-
+	s := rand.NewSource(time.Now().UnixNano())
+	r := rand.New(s)
 	timeoutC := time.After(timeout)
 	for {
 		// Perform as many queries as possible within this time.
 		select {
 		case <-timeoutC:
-			// Calculate averages on completion.
-			resultStorage.lock.Lock()
-			resultStorage.TotalDuration = totalDuration.Seconds()
-			if resultStorage.Successes > 0 {
-				resultStorage.AverageDuration = totalDuration.Seconds() / float64(resultStorage.Successes)
-				resultStorage.QueriesPerSecond = 1 / resultStorage.AverageDuration
-			}
-			resultStorage.lock.Unlock()
-
 			return
 		default:
 		}
 
+		n := r.Intn(len(totals.results))
+		resultStorage := totals.results[n]
 		// Drop out if we have too many errors.
-		if resultStorage.GetErrors() >= errorThreshold {
-			level.Error(logger).Log("too many errors, skipping the remainder", resultStorage.Query)
-			return
+		if totals.getErrors() >= *fErrorThreshold {
+			level.Info(logger).Log("msg", fmt.Sprintf("too many errors, skipping the remainder %s", resultStorage.Query))
 		}
 
-		// Start timer.
 		queryStart := time.Now()
+		resp, err := http.Get(resultStorage.queryURL.String())
+		duration := time.Since(queryStart)
 
-		// Run query.
-		resp, err := http.Get(query)
 		if err != nil || resp.StatusCode != 200 {
-			resultStorage.AddError()
+			totals.addError()
```

Comment:
you have to read body in case of resp.StatusCode, otherwise you are leaking go routines and you will can potentially get throttled on large number of TCP connectiontions. Plus it can affect benchmarks.

Reply:
still applies